### PR TITLE
Fix for https://github.com/OWASP/owasp-masvs/issues/203

### PR DESCRIPTION
**Pull request by commjoen**
Thank you for submitting a Pull Request to the Mobile Security Testing Guide. Please make sure that:

- [ ] Your contribution is written in the 2nd person (e.g. you)
- [ ] Your contribution is written in an active present form for as much as possible.
- [ ] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [ ] Your contribution has proper formatted markdown and/or code
- [ ] Any references to website have been formatted as [TEXT](URL “NAME”)
- [ ] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

If your PR is related to an issue. Please end your PR test with the following line:
This PR covers issue https://github.com/OWASP/owasp-masvs/issues/203.

### CODE REVIEW

File: Document/0x06f-Testing-Local-Authentication.md [hunk 1]
```diff
@@ -198,6 +198,8 @@ It is important to remember that Local Authentication framework is an event-base
 - Verify that sensitive processes, such as re-authenticating a user triggering a payment transaction, are protected using the Keychain services method.
 - Verify that the `kSecAccessControlTouchIDAny` or `kSecAccessControlTouchIDCurrentSet` flags are set and `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` protection classes are set when the `SecAccessControlCreateWithFlags` method is called. Note that, alternatively, `kSecAccessControlUserPresence` can be used as a flag as well when you want to be able to use passcode as a fallback. Last, note that, when `kSecAccessControlTouchIDCurrentSet` is set, changing the fingerprints registered to the device will invalidate the entry which is protected with the flag.
 
+Please note that using `kSecAccessControlTouchIDAny` will allow the keymaterial to be used when the user (re-)enroles new fingerprints, but it will mean that attackers who are somehow able to enrole their fingerprints to the device, now can access items that use this flag as well.
```
Comment (cpholguera) [inline Document/0x06f-Testing-Local-Authentication.md]: Keep it consistent by referring to "entries". Split and made the sentences shorter.

```suggestion
Please note that using `kSecAccessControlTouchIDAny` will allow the entry to be used when the user (re-)enroles new fingerprints. However, it means also that attackers, who are somehow able to enrole their fingerprints to the device, can now access those entries as well.
```

File: Document/0x06f-Testing-Local-Authentication.md [hunk 2]
```diff
@@ -198,6 +198,8 @@ It is important to remember that Local Authentication framework is an event-base
 - Verify that sensitive processes, such as re-authenticating a user triggering a payment transaction, are protected using the Keychain services method.
 - Verify that the `kSecAccessControlTouchIDAny` or `kSecAccessControlTouchIDCurrentSet` flags are set and `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` protection classes are set when the `SecAccessControlCreateWithFlags` method is called. Note that, alternatively, `kSecAccessControlUserPresence` can be used as a flag as well when you want to be able to use passcode as a fallback. Last, note that, when `kSecAccessControlTouchIDCurrentSet` is set, changing the fingerprints registered to the device will invalidate the entry which is protected with the flag.
```
Comment (cpholguera) [inline Document/0x06f-Testing-Local-Authentication.md]: Some remarks:
- I can't find an explanation on why we should test for kSecAccessControlTouchIDAny (or the other flags) being set
- kSecAccessControlTouchIDAny is since iOS 11.3 deprecated, use kSecAccessControlBiometryAny instead.

**Comment by OMTGreleaser** [comment]
Results for commit 1103f4e7ea4200c998071062868837ffa7e51c8f: Broken link result: [!] Error(s) found by the Links validator: 9 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md (hunk 1 above)
Comment (TheDauntless) [reply Document/0x06f-Testing-Local-Authentication.md]: Modified your comment for a grammar improvement since I can't propose a new commit suggestion on top of yours.

**Comment by OMTGreleaser** [comment]
Results for commit 8f86bd407de1f9f40456d127101bd46c0bc4e904: Broken link result: [!] Error(s) found by the Links validator: 10 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 0406304436ec7f8cae63b5839d3cd3ba99e3db61: Broken link result: [!] Error(s) found by the Links validator: 10 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 0fbe901a0cc57a625d439e54a841f10227d4f480: Broken link result: [!] Error(s) found by the Links validator: 10 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md (hunk 2 above)
Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: Thanks! Updated it :).

**Comment by OMTGreleaser** [comment]
Results for commit f895c2313ca9c7f79eece329a7a38d4cdaf0176e: Broken link result: [!] Error(s) found by the Links validator: 12 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md [hunk 3]
```diff
@@ -193,10 +193,14 @@ If `Security.framework` is used, only the second one will be shown.
...
+  - `kSecAccessControlTouchIDCurrentSet`(since iOS 11.3 `kSecAccessControlBiometryCurrentSet`). This will make sure that a user needs to authenticate with his biometrics (e.g. Face ID or Touch ID)before he can access the data in the Keychain entry. Whenever the user adds a fingerprint or facial representation to the device, it will automatically invalidate the entry in the Keychain. This makes sure that the Keychain-entry is protected against adversarial additions to the Keychain.
+  - `kSecAccessControlTouchIDAny` (since iOS 11.3 `kSecAccessControlBiometryAny`). This will make sure that a user needs to authenticate with his biometrics (e.g. Face ID or Touch ID)before he can access the data in the Keychain entry. The Keychain-entry will survive any (re-)enroling of new fingerprints or facial representation. This can be very practical if the user has a changing fingerprint (which happens in case of lumberjackers for instance). However, it also means that attackers, who are somehow able to enrole their fingerprints or facial representations to the device, can now access those entries as well.
```
Comment (cpholguera) [inline Document/0x06f-Testing-Local-Authentication.md]:
```suggestion
  - `kSecAccessControlTouchIDAny` (since iOS 11.3 `kSecAccessControlBiometryAny`). This will make sure that a user needs to authenticate with his biometrics (e.g. Face ID or Touch ID) before he can access the data in the Keychain entry. The Keychain entry will survive any (re-)enroling of new fingerprints or facial representation. This can be very practical if the user has a changing fingerprint. However, it also means that attackers, who are somehow able to enrole their fingerprints or facial representations to the device, can now access those entries as well.
```

Comment (TheDauntless) [inline Document/0x06f-Testing-Local-Authentication.md]:
```suggestion
  - `kSecAccessControlBiometryAny` (before iOS 11.3 `kSecAccessControlTouchIDAny`). This will make sure that a user needs to authenticate with his biometrics (e.g. Face ID or Touch ID) before he can access the data in the Keychain entry. The Keychain-entry will survive any (re-)enrolling of new fingerprints or appearances. This can prevent negative feedback from users who are disgruntled that updating their Face ID or Touch ID settings forced them to re-enrol into the application. However, it also means that attackers, who are somehow able to enrol their fingerprints or appearance to the device, can now access those entries as well.
```

File: Document/0x06f-Testing-Local-Authentication.md [hunk 4]
```diff
@@ -193,10 +193,14 @@ If `Security.framework` is used, only the second one will be shown.
...
+  - `kSecAccessControlUserPresence` can be used as an alternative. This will allow the user to authenticate himself through a PassCode if the biometric authentication no longer works.
```
Comment (TheDauntless) [inline Document/0x06f-Testing-Local-Authentication.md]:
```suggestion
  - `kSecAccessControlUserPresence` can be used as an alternative. This will allow the user to authenticate himself through a PassCode if the biometric authentication no longer works. This is considered to be weaker than kSecAccessControlBiometryAny since it is much easier to steal someone's pincode by seeing them enter it, than it is to bypass the Touch ID or Face ID service.
```

File: Document/0x06f-Testing-Local-Authentication.md [hunk 5]
```diff
@@ -193,10 +193,14 @@ If `Security.framework` is used, only the second one will be shown.
...
+- Verify that `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` protection class is set when the `SecAccessControlCreateWithFlags` method is called to ensure that the device is at least protected with a passcode and that the keychain entries can therefore be secured.
```
Comment (TheDauntless) [inline Document/0x06f-Testing-Local-Authentication.md]: The previous stuff is explanation on the flags, while this is a recommendation. You als can't have Touch ID or Face ID without having a pincode. I would rephrase to fit the style of the previous bullets and explain when you can use it.

File: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 6]
```diff
@@ -579,7 +579,7 @@ Before we describe the usable identifiers, let's quickly discuss how they can be
...
+Any scheme based on these methods will be more secure the moment a passcode and/or touch-id is enabled, the materials stored in the Keychain or filesystem are protected with protection classes (such as `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly` and `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`), and the `SecAccessControlCreateFlags` is set either with `kSecAccessControlDevicePasscode` (for passcodes), `kSecAccessControlUserPresence` (passcode, Face ID or Touch ID), `kSecAccessControlTouchIDAny` / `kSecAccessControlBiometryAny` (Touch ID / Face ID) or `kSecAccessControlTouchIDCurrentSet` / `kSecAccessControlBiometryCurrentSet` (Face ID / Touch ID: but current enroled biometrics only).
```
Comment (TheDauntless) [inline Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md]:
```suggestion
Any scheme based on these methods will be more secure the moment a passcode and/or Touch ID is enabled, the materials stored in the Keychain or filesystem are protected with protection classes (such as `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly` and `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`), and the `SecAccessControlCreateFlags` is set either with `kSecAccessControlDevicePasscode` (for passcodes), `kSecAccessControlUserPresence` (passcode, Face ID or Touch ID), kSecAccessControlTouchIDAny  (Touch ID), `kSecAccessControlBiometryAny` (Touch ID / Face ID) or `kSecAccessControlTouchIDCurrentSet` / `kSecAccessControlBiometryCurrentSet` (Face ID / Touch ID: but current enrolled biometrics only).
```

File: Document/0x06f-Testing-Local-Authentication.md (hunk 3 above)
Comment (cpholguera) [reply Document/0x06f-Testing-Local-Authentication.md]: Some corrections and pun removed. I hope that's ok.

File: Document/0x06f-Testing-Local-Authentication.md (hunk 4 above)
Comment (cpholguera) [inline Document/0x06f-Testing-Local-Authentication.md]:
```suggestion
  - `kSecAccessControlUserPresence` can be used as an alternative. This will allow the user to authenticate through a passcode if the biometric authentication no longer works.
```

File: Document/0x06f-Testing-Local-Authentication.md (hunk 5 above)
Comment (cpholguera) [inline Document/0x06f-Testing-Local-Authentication.md]:
```suggestion
- Verify that `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` protection class is set when the `SecAccessControlCreateWithFlags` method is called to ensure that the device is at least protected with a passcode and that the Keychain entries can therefore be secured.
```

Comment (cpholguera) [reply Document/0x06f-Testing-Local-Authentication.md]: Have we at this point already explained what "protection classes" are ?

File: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md (hunk 6 above)
Comment (cpholguera) [inline Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md]:
```suggestion
Any scheme based on these methods will be more secure the moment a passcode and/or Touch ID is enabled, the materials stored in the Keychain or filesystem are protected with protection classes (such as `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly` and `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`), and the `SecAccessControlCreateFlags` is set either with `kSecAccessControlDevicePasscode` (for passcodes), `kSecAccessControlUserPresence` (passcode, Face ID or Touch ID), `kSecAccessControlTouchIDAny` / `kSecAccessControlBiometryAny` (Touch ID / Face ID) or `kSecAccessControlTouchIDCurrentSet` / `kSecAccessControlBiometryCurrentSet` (Face ID / Touch ID: but current enroled biometrics only).
```

File: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 7]
```diff
@@ -579,7 +579,7 @@ Before we describe the usable identifiers, let's quickly discuss how they can be
 - You can store something in the Keychain to identify the application's instance. To make sure that this data is not backed up, use `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` (if you want to secure the data and properly enforce a passcode or touch-id requirement), `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`, or `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`.
```
Comment (cpholguera) [inline Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md]:
```suggestion
- You can store something in the Keychain to identify the application's instance. To make sure that this data is not backed up, use `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` (if you want to secure the data and properly enforce a passcode or Touch ID requirement), `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`, or `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`.
```

Comment (cpholguera) [reply Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md]: Please revise the consistency of Touch ID and Face ID in this file and in 0x6f at least. Thank you :)

File: Document/0x06f-Testing-Local-Authentication.md [hunk 8]
```diff
@@ -193,10 +193,14 @@ If `Security.framework` is used, only the second one will be shown.
...
+- Verify that access control flags are set for the keychain entry which ensure that the keychain item its data can only be unlocked by means of authenticating the user. This can be with one of the following flags:
+  - `kSecAccessControlTouchIDCurrentSet`(since iOS 11.3 `kSecAccessControlBiometryCurrentSet`). This will make sure that a user needs to authenticate with his biometrics (e.g. Face ID or Touch ID)before he can access the data in the Keychain entry. Whenever the user adds a fingerprint or facial representation to the device, it will automatically invalidate the entry in the Keychain. This makes sure that the Keychain-entry is protected against adversarial additions to the Keychain.
```
[the remark this answers is not in the document]
Comment (cpholguera) [reply Document/0x06f-Testing-Local-Authentication.md]: I'd rather give preference to the new stuff. I'd switch the first part and write (before iOS 11.3 `...TouchID...`)

File: Document/0x06f-Testing-Local-Authentication.md (hunk 3 above)
Comment (cpholguera) [reply Document/0x06f-Testing-Local-Authentication.md]: I'd rather give preference to the new stuff. I'd switch the first part and write (before iOS 11.3 `...TouchID...`)

**Comment by OMTGreleaser** [comment]
Results for commit e128865f2c4a13fa33584b0e95967d90480c8530: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md (hunk 3 above)
Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: > Some corrections and pun removed. I hope that's ok.

well it was no pun, it is actually a known fact that people working with wood have this problem. But will keep that out then :).

**Comment by OMTGreleaser** [comment]
Results for commit ea9b525a498c2502d8ce058d27c9af2ceb6993a4: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md (hunk 5 above)
Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: They're both recomendations to me though , but i will rewrite it a bit ;-)

**Comment by OMTGreleaser** [comment]
Results for commit e7d124276123be2909099dcd4e69b4fe7d38055a: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit b7789991f0b87684d3d656f16b556f71b5151626: Broken link result: [!] Error(s) found by the Links validator: 8 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md (hunk 5 above)
Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: I did now, thanks!

File: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md (hunk 7 above)
Comment (commjoen) [reply Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md]: Done :D thanks!

**Comment by OMTGreleaser** [comment]
Results for commit b15957fa26d5b201dda6b1bf0af5868561ce5324: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 7001ab446fa64da5b89deaefca7a1b1194ea7ba2: Broken link result: [!] Error(s) found by the Links validator: 6 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06f-Testing-Local-Authentication.md [hunk 9]
```diff
@@ -126,7 +126,7 @@ if status == noErr {
     }
 ```
 
-Now we can request the saved item from the Keychain. Keychain Services will present the authentication dialog to the user and return data or nil depending on whether a suitable fingerprint was provided or not.
+Now we can request the saved item from the keychain. keychain services will present the authentication dialog to the user and return data or nil depending on whether a suitable fingerprint was provided or not.
```
Comment (cpholguera) [inline Document/0x06f-Testing-Local-Authentication.md]: Uppercase here because it starts the sentence

```suggestion
Now we can request the saved item from the keychain. Keychain services will present the authentication dialog to the user and return data or nil depending on whether a suitable fingerprint was provided or not.
```

**Comment by OMTGreleaser** [comment]
Results for commit f720626c24fd5ded1294040733f727b683302219: Broken link result: [!] Error(s) found by the Links validator: 7 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        0 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.
